### PR TITLE
Populate VCH status on VIC admin page

### LINKED ISSUE #1229: Add "healthy" setup checks into vic-admin web page.
I recently setup a new ESX and forgot to change my firewall rules so VIC was not working properly. I would be nice to be able to go to the vicadmin web page and get a "red/yellow/green" indicator of the setup of VIC in that environment. I'm thinking of something where it shows if all the hosts have the firewall rules set properly, the available space in the datastore being used, ability to connect to public repos, etc.. 

So in my case: If I'd gone to the vicadmin page to pull logs because attach was not working I would have seen an ESX listed as not having a correct firewall configuration and I would have known my issue was my configuration. 

Having this information would also be useful for the log bundles to provide context of how "healthy" the environment is at the time.

### PR DESCRIPTION
**Pull request by kjplatz**
Fixes #1229

This populates the "VCH Status" health check on the VICadmin page by verifying that "vic-init" and all proceses defined in the VirtualContainerHostConfigSpec.ExecutorConfig are currently running.  If any processes are not running, the green check changes to an orange warning triangle, and a warning message indicating that that particular process is not running is published on the vicadmin page.

### CODE REVIEW

File: lib/tether/tether.go [hunk 1]
```diff
@@ -133,6 +136,15 @@ func (t *tether) setup() error {
 		}
 	}
 
+	// Create PID file for tether
+	paths := strings.Split(os.Args[0], "/")
+	err = ioutil.WriteFile(fmt.Sprintf("%s/%s.pid", PIDFileDir, paths[len(paths)-1]),
```
Comment (hmahmood) [inline lib/tether/tether.go]: There is no need to create a pid file for the tether. You should only create pid files for the processes the tether launches.

File: lib/tether/tether.go [hunk 2]
```diff
@@ -39,6 +41,7 @@ const (
 
 	// the length of a truncated ID for use as hostname
 	shortLen = 12
+	PIDFileDir = "/var/run"
```
Comment (hmahmood) [inline lib/tether/tether.go]: This should be in tether_linux.go .

File: lib/tether/tether.go [hunk 3]
```diff
@@ -305,6 +317,10 @@ func (t *tether) handleSessionExit(session *SessionConfig) {
 		logs = logs[logCount-MaxDeathRecords+1:]
 	}
 
+	// Remove associated PID file
+	paths := strings.Split(session.Cmd.Path, "/")
+	_ = os.Remove(fmt.Sprintf("%s/%s.pid", PIDFileDir, paths[len(paths)-1]))
```
Comment (hmahmood) [inline lib/tether/tether.go]: If there is no `/` in the path, paths will be empty and `len(paths)-1` will be `-1`.

File: lib/tether/tether.go [hunk 4]
```diff
@@ -305,6 +317,10 @@ func (t *tether) handleSessionExit(session *SessionConfig) {
 		logs = logs[logCount-MaxDeathRecords+1:]
 	}
 
+	// Remove associated PID file
+	paths := strings.Split(session.Cmd.Path, "/")
```
Comment (hmahmood) [inline lib/tether/tether.go]: You can use [path.Base](https://golang.org/pkg/path/#Base)

File: lib/tether/tether.go (hunk 3 above)
Comment (hmahmood) [inline lib/tether/tether.go]: ... and [path.Join](https://golang.org/pkg/path/#Join) here

File: lib/vicadmin/validate.go [hunk 5]
```diff
@@ -191,3 +202,55 @@ func (v *Validator) QueryDatastore(ctx context.Context, vch *config.VirtualConta
...
+
+	procs := []string { "vic-init" }
+
+	// Extract required components from vchConfig
+	// Only report on components with Restart set to true
+	for _, sess := range vch.ExecutorConfig.Sessions {
+		if !sess.Restart {
+			continue
+		}
+		parsedcmd := basename.FindStringSubmatch(sess.Cmd.Path)
```
Comment (hmahmood) [inline lib/vicadmin/validate.go]: Maybe you need to write a helper function that return the pid file path given a session object. That way you can consistently use it everywhere in the tether and vidadmin.

Comment (hmahmood) [inline lib/vicadmin/validate.go]: A pidfile type/package perhaps?

File: lib/vicadmin/validate.go [hunk 6]
```diff
@@ -191,3 +202,55 @@ func (v *Validator) QueryDatastore(ctx context.Context, vch *config.VirtualConta
...
+		// Field 3 is the current state as reported by the kernel
+		switch fields[2][0] {
+		// We're good
+		case 'R', 'S', 'D': break
+		// Process has been killed, is dying, or a zombie
+		case 'K', 'X', 'x', 'Z':
+			v.VCHIssues = template.HTML(fmt.Sprintf("%s<span class=\"error-message\">Process %s has failed</span>\n",
+				v.VCHIssues, err))
+		}
+		log.Infof("Process %s running as PID %s", proc, pid)
```
Comment (hmahmood) [inline lib/vicadmin/validate.go]: We are going to fall through here in either case from above; is this message accurate?

File: lib/vicadmin/validate.go [hunk 7]
```diff
@@ -191,3 +202,55 @@ func (v *Validator) QueryDatastore(ctx context.Context, vch *config.VirtualConta
...
+		}
+
+		fields := strings.Split(string(status), " ")
+		// Field 3 is the current state as reported by the kernel
+		switch fields[2][0] {
+		// We're good
+		case 'R', 'S', 'D': break
+		// Process has been killed, is dying, or a zombie
+		case 'K', 'X', 'x', 'Z':
+			v.VCHIssues = template.HTML(fmt.Sprintf("%s<span class=\"error-message\">Process %s has failed</span>\n",
```
Comment (hmahmood) [inline lib/vicadmin/validate.go]: If this is going to be displayed to the user, I would use the name of the service whose process if failing.

File: lib/tether/tether.go (hunk 1 above)
Comment (kjplatz) [inline lib/tether/tether.go]: @hmahmood Per our discussion, we may be moving vicadmin to be its own service not spawned by the tether.  In that event, it does make sense to create a PID file for the tether.

**Comment by hmahmood** [comment]
lgtm

**Comment by matthewavery** [comment]
lgtm

**Comment by kjplatz** [comment]
Bing

**Pull request merged**